napari/layers/utils/property_map.py

```diff
@@ -0,0 +1,210 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Generic, Iterable, List, TypeVar
+
+from pydantic import validator
+
+from ...utils import Colormap
+from ...utils.colormaps import ValidColormapArg, ensure_colormap
+from ...utils.events import EventedModel
+from ...utils.events.custom_types import Array
+from .color_transformations import ColorType
+
+OutputType = TypeVar('OutputType')
+
+
+class PropertyMap(ABC, Generic[OutputType], EventedModel):
+    values: List[OutputType] = []
```

Comment:
i wonder whether this might make more sense as a [cached_property](https://docs.python.org/3/library/functools.html#functools.cached_property)?  It seems to be more computed _from_ the model based on some input (properties), rather than an attribute of the model itself?

In that case, this ABC needn't inherit from `BaseModel` at all...

also, whose responsibility is it to call `refresh`?  does this happen greedily?  or could one clear the cache instead, and let it be recalculated only when necessary?

Reply:
It's my first time seeing `@cached_property`, but based on the Python docs it doesn't seem to fit here because `properties` is mutable input rather than immutable instance state.

The mutability of `properties` is why we need `refresh`, `add`, and `remove`. And as a first pass, I suggest making it the responsibility of the owner (i.e. Layer) to call those, with some public way in `Layer` for the user to force refresh everything. Lazy evaluation could be useful if we change the way that the indices of the elements in view are passed through from a vispy layer to a layer, but that requires a little more thought.

I could lean towards making `values` a read-only property in `PropertyMap`. But then it (or something similar) needs to be writable in `DirectPropertyMap` to support the current GUI interactions associated with that coloring mode.